ctats/bus.go

```diff
@@ -0,0 +1,109 @@
+package ctats
+
+import (
+	"context"
+
+	"github.com/puzpuzpuz/xsync/v3"
+)
+
+type metricsBusKey string
+
+const defaultCtxKey metricsBusKey = "default_metrics_bus_key"
+
+func fromCtx(ctx context.Context) *bus {
+	if ctx == nil {
+		return nil
+	}
+
+	dn := ctx.Value(defaultCtxKey)
+
+	if dn == nil {
+		return nil
+	}
+
+	return dn.(*bus)
+}
+
+func embedInCtx(ctx context.Context, b *bus) context.Context {
+	return context.WithValue(ctx, defaultCtxKey, b)
+}
+
+type bus struct {
+	counters   *xsync.MapOf[string, adder]
+	gauges     *xsync.MapOf[string, recorder]
+	histograms *xsync.MapOf[string, recorder]
+
+	// initializedToNoop is a testing convenience flag that identifies
+	// whether the OTEL client should be configured or not.
+	initializedToNoop bool
+}
+
+func newBus() *bus {
+	return &bus{
+		counters:   xsync.NewMapOf[string, adder](),
+		gauges:     xsync.NewMapOf[string, recorder](),
+		histograms: xsync.NewMapOf[string, recorder](),
+	}
+}
+
+// counterFromCtx retrieves the counter instance from the metrics bus
+// in the context.  If the ctx has no metrics bus, or if the bus does
+// not have a counter for the provided ID, returns nil.
+func (b bus) getCounter(
+	id string,
+) adder {
+	if b.initializedToNoop {
+		c, _ := b.counters.LoadOrStore(formatID(id), &noopAdder{})
+		return c
+	}
+
+	if b.counters == nil {
+		return nil
+	}
+
+	c, _ := b.counters.Load(formatID(id))
+
+	return c
+}
+
+// gaugeFromCtx retrieves the gauge instance from the metrics bus
+// in the context.  If the ctx has no metrics bus, or if the bus does
+// not have a gauge for the provided ID, returns nil.
+func (b bus) getGauge(
+	ctx context.Context,
+	id string,
+) recorder {
+	if b.initializedToNoop {
+		g, _ := b.gauges.LoadOrStore(formatID(id), &noopRecorder{})
+		return g
+	}
+
+	if b.counters == nil {
+		return nil
+	}
+
+	g, _ := b.gauges.Load(formatID(id))
+
+	return g
+}
+
+// histogramFromCtx retrieves the histogram instance from the metrics bus
+// in the context.  If the ctx has no metrics bus, or if the bus does
+// not have a histogram for the provided ID, returns nil.
+func (b bus) getHistogram(
+	ctx context.Context,
+	id string,
+) recorder {
+	if b.initializedToNoop {
+		h, _ := b.histograms.LoadOrStore(formatID(id), &noopRecorder{})
```

Comment:
heh it's times like these that I almost miss C++ templating...

Reply:
I wouldn't mind that myself.